pkg/kn/commands/service/service_create.go

```diff
@@ -53,66 +60,185 @@ func NewServiceCreateCommand(p *commands.KnParams) *cobra.Command {
+	_, err := client.Services(namespace).Get(name, v1.GetOptions{})
+	if api_errors.IsNotFound(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// Create service struct from provided options
+func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name string, namespace string) (*serving_v1alpha1_api.Service,
+	error) {
+
+	service := serving_v1alpha1_api.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+
+	// TODO: Should it always be `runLatest` ?
+	service.Spec.DeprecatedRunLatest = &serving_v1alpha1_api.RunLatestType{
+		Configuration: serving_v1alpha1_api.ConfigurationSpec{
+			DeprecatedRevisionTemplate: &serving_v1alpha1_api.RevisionTemplateSpec{
+				Spec: serving_v1alpha1_api.RevisionSpec{
+					DeprecatedContainer: &corev1.Container{},
+				},
+			},
+		},
+	}
+
+	err := editFlags.Apply(&service, cmd)
+	if err != nil {
+		return nil, err
+	}
+	return &service, nil
+}
+
+func showUrl(client serving_v1alpha1_client.ServingV1alpha1Interface, serviceName string, namespace string, out io.Writer) error {
+	service, err := client.Services(namespace).Get(serviceName, v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot fetch service %s in namespace %s for extracting the URL", serviceName, namespace)
+	}
+	url := service.Status.URL.String()
+	if url == "" {
+		url = service.Status.DeprecatedDomain
+	}
+	fmt.Fprintln(out, "\nService URL:")
+	fmt.Fprintf(out, "%s\n", url)
+	return nil
+}
+
+// Duck type for writers having a flush
+type flusher interface {
+	Flush() error
+}
+
+func waitForService(client serving_v1alpha1_client.ServingV1alpha1Interface, serviceName string, namespace string, out io.Writer, timeout int) error {
+	// Wait for service to enter 'Ready' state, with a timeout of which is slightly larger than the provided
+	// timeout. We have our own timeout which fire after "timeout" seconds and stop the watch
+	var timeoutWatch int64 = int64(timeout + 30)
+	opts := v1.ListOptions{
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", serviceName).String(),
+		TimeoutSeconds: &timeoutWatch,
+	}
+	watcher, err := client.Services(namespace).Watch(opts)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(out, "Waiting for service %s to become ready ... ", serviceName)
+	if flusher, ok := out.(flusher); ok {
+		flusher.Flush()
+	}
+
+	if err := waitForReadyCondition(watcher, timeout); err != nil {
+		fmt.Fprintln(out)
+		return err
+	}
+	fmt.Fprintln(out, "OK")
+	return nil
+}
+
+func waitForReadyCondition(watcher watch.Interface, timeout int) error {
+	defer watcher.Stop()
+	for {
+		select {
+		case <-time.After(time.Duration(timeout) * time.Second):
+			return fmt.Errorf("timeout: Service not ready after %d seconds", timeout)
+		case event, ok := <-watcher.ResultChan():
+			if !ok || event.Object == nil {
+				return errors.New("timeout while waiting for service to become ready")
```

Comment:
I guess we to deal with the `unknown` status of `Ready` the idea would be to not timeout here but instead reset the timeout clock, giving a chance to the service to become ready from a `unknown` state. This would be for every case of `true` to `unknown` transition. Thoughts?

Reply:
I think it's ok to complicate the code in that way later and not now.

Reply:
Good idea, @maximilien but I agree with @sixolet to keep simple for now, refactor to be reusable for other resources and the open an issue for this particular feature (i.e. don't timeout as long as in `Unknown` state)

Reply:
OK fair enough. Cool.